tls/s2n_connection.c

```diff
@@ -1307,6 +1307,19 @@ static S2N_RESULT s2n_connection_set_max_fragment_length(struct s2n_connection *
         conn->max_outgoing_fragment_length = max_frag_length;
     }
 
+    /* If no buffer has been initialized yet, no need to resize.
+     * The standard I/O logic will handle initializing the buffer.
+     */
+    if (s2n_stuffer_is_empty(&conn->out)) {
+        return S2N_RESULT_OK;
+    }
+
+    uint16_t max_wire_record_size = 0;
+    RESULT_GUARD(s2n_record_max_write_record_size(conn, conn->max_outgoing_fragment_length, &max_wire_record_size));
+    if ((conn->out.blob.size < max_wire_record_size)) {
+        RESULT_GUARD_POSIX(s2n_realloc(&conn->out.blob, max_wire_record_size));
```

Comment:
What if we an application does this?

```
struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
s2n_connection_prefer_throughput(conn);
/* Do stuff */
...
/* Done, restore the conn for a fresh session. */
s2n_connection_wipe(conn);
s2n_connection_prefer_throughput(conn);
```

Will s2n reallocate the buffer after the call to s2n_connection_prefer_throughput after calling s2n_connection_wipe? Current behavior is that it will not, which avoids unnecessary allocation/de-allocation. I think it's worth making sure we preserve the current behavior, if possible.

Reply:
IGood news / bad news situation: this isn't a problem because we already free the I/O buffer on wipe and have to do those allocations: https://github.com/aws/s2n-tls/blob/main/tls/s2n_connection.c#L669-L672 @baldwinmatt made that change almost two years ago.

Basically, the output buffers are always zero to start and are resized when the first message is sent.

Reply:
Unless I'm missing something, @baldwinmatt modified the s2n_connection_wipe behavior two years ago to always free the I/O memory: https://github.com/aws/s2n-tls/blob/main/tls/s2n_connection.c#L669-L672 The buffer are allocated when the first message is received

Reply:
Since this is important to this change, I added a test to verify. We definitely do free the I/O memory on s2n_connection_wipe.